### DISCUSSION TITLE
magit-process-interrupt

### DISCUSSION THREAD

**Discussion by petergaultney** [magit/magit#5086] [Ideas]
`magit-process-interrupt` works, in that it does successfully kill the running process. However, this is not ergonomic when using some kinds of pre-commit hooks, like for instance [`pre-commit`](https://pre-commit.com/).

Whatever signal `magit` is sending is an abrupt process kill, the result of which is that a patch file that pre-commit has created to exclude files from its checks never gets reapplied to the repository. This means that changes 'appear' to be lost, when in fact they're just hiding. Some people will know how to have their changes restored from the patch file (and that those patch files live in the user's home directory), but others will not.

When running `git` at the shell, in constrast, the standard procedure for a user who wants to stop the commit process is to press Ctrl-C, which sends `SIGINT` to the process, which `pre-commit` handles gracefully, restoring the patch file and leaving things the way a user would expect.

Supporting a new command, `magit-process-interrupt` (possibly aliased as `magit-process-ctrl-c`?) would be a nice way to offer users an alternative that does "what they expect" when dealing with long-running pre-commit hooks (and probably other use cases that I don't personally run into).

Does this sound like something the project would be interested in supporting? It seems likely to be relatively trivial[^1], and I'm willing to take a stab at it myself even though I'm no elisp guru.

[^1]: ~i have now looked around to see whether the underlying Emacs `kill-process` function has an analogue for sending SIGINT, and so far I'm coming up empty. Possibly this will not be as trivial as I hoped. :)~ it looks like `interrupt-process` is a built in function and does send a SIGINT. https://www.gnu.org/software/emacs/manual/html_node/elisp/Signals-to-Processes.html

**Comment by petergaultney** [comment]
here's a potential start on a PR. it looks like this should not be terribly difficult to accomplish.

**Comment by kyleam** [comment]
Here's the approach proposed by @tarsius in <https://github.com/magit/magit/pull/5087#issuecomment-1947852173>:

 > We could change `magit-process-kill` to first send `SIGINT` and remember that using `process-put`, and then `SIGKILL` on the second try. The user wouldn't even have to know about this, a thought process like "hm, somehow it didn't work [maybe I pressed the wrong key?], let's try again" would be enough.

**Reply by tarsius** [reply]
Implemented in 5e176cf1063aa67e400a67f7d773a4175f107dd1.
